### ISSUE TITLE
Crash on WebsocketClient.ReceiveAsync with Memory<T>

### ISSUE THREAD

**Issue by nmoschkin** [dotnet/runtime#59362]
### Describe the bug

Using the WebsocketClient.ReceiveAsync with a Memory<byte> buffer causes an exception:

```
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
```

When executing synchronously with ConfigureAwait inside of a non-async method.  

The method is its own thread that was created with System.Threading.Thread with IsBackground = true.

This same method works if I use an ArraySegment instead of Memory.  It was also working in the previous .NET 6.0 preview _with_ the Memory<T>.

### To Reproduce
```c#
            byte[] inputChunk = new byte[1024];

            ValueWebSocketReceiveResult result;
            var memTarget = new Memory<byte>(inputChunk);

            // loop forever or until the connection is broken or canceled.
            while (!ctsReceive.IsCancellationRequested && socket?.State == WebSocketState.Open)
            {
                try
                {
                    result = socket.ReceiveAsync(memTarget, ctsReceive.Token)
                        .ConfigureAwait(false)
                        .GetAwaiter()
                        .GetResult();
                }
                catch (Exception ex)
                {
                    Console.Clear();
                    Console.WriteLine(ex.ToString());

                    Environment.Exit(0);
                    return;
                }
            }

```
### Exceptions (if any)
```
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
```

### Further technical details
- Include the output of `dotnet --info`

```
.NET SDK (reflecting any global.json):
 Version:   6.0.100-rc.1.21458.32
 Commit:    d7c22323c4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-rc.1.21458.32\

Host (useful for support):
  Version: 6.0.0-rc.1.21451.13
  Commit:  d7619cd4b1

.NET SDKs installed:
  3.1.410 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]
  5.0.104 [C:\Program Files\dotnet\sdk]
  5.0.201 [C:\Program Files\dotnet\sdk]
  5.0.204 [C:\Program Files\dotnet\sdk]
  5.0.301 [C:\Program Files\dotnet\sdk]
  5.0.401 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.3.21202.5 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.6.21355.2 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.1.21458.32 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.6.21355.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.1.21452.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.6.21352.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.1.21451.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-preview.6.21353.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.1.21451.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version

Visual Studio 2022 Preview 4.

Windows 11 22000.184

**Comment by CarnaViire** [comment]
Thanks for reporting.
I can reproduce the issue.

**Comment by CarnaViire** [comment]
I confirm the regression was introduced by https://github.com/dotnet/runtime/pull/56282
/cc @stephentoub

**Comment by stephentoub** [comment]
This is expected.  The method returns `ValueTask<T>`, and using `GetAwaiter().GetResult()` on a `ValueTask<T>` is not supported unless you know the `ValueTask<T>` to have already completed.  It happened to previously work because the `ValueTask<T>` wrapped a `Task<T>`, but now it doesn't.  If you enable the CA2012 static analysis rule, it should flag this misuse.

You can fix your code by adding a `.AsTask()` call in before the `.GetAwaiter()` call.

**Comment by nmoschkin** [comment]
Okay.  I read the entire article, and that certainly clarified things.  I have just never run across this, yet, and I've run across a lot of things. ... well this is certainly expositive.  Thanks for your time.

**Comment by MihaZupan** [comment]
> I was trying to get more performance out of that code path

What is the reason you are not using `async`/`await`? Perf-wise, blocking waits are a problem.

**Comment by nmoschkin** [comment]
I think I was having trouble starting an async thread that wasn't immediately passing control to another thread.  I suppose I was asking for too much.

**Comment by nmoschkin** [comment]
You know that's a good question.  I haven't modified this code in a very long time, and many other things have changed since then.  

I think I wanted the neatness of being able to just pass a function as a new System.Threading.Thread, and I was concerned that starting an async thread would possibly just pass the control on to another thread, immediately, completely defeating the purpose of why I had created the thread in the first place.  

I guess a good question to ask would be when you start an async thread and then await a function that never returns, are you creating a thread just to jump to a new thread that's running with a thread join pattern?  Or does it somehow know NOT to start a new thread because it just started a new thread?

**Comment by nmoschkin** [comment]
> This is expected / by design and the repro is buggy even though it happened to previously "work".
> 
> The method returns `ValueTask<T>`, and using `GetAwaiter().GetResult()` on a `ValueTask<T>` is not supported unless you know the `ValueTask<T>` to have already completed. It happened to previously work because the `ValueTask<T>` wrapped a `Task<T>`, but now it doesn't. If you enable the CA2012 static analysis rule, it should flag this misuse.
> 
> You can fix your code by adding a `.AsTask()` call in before the `.GetAwaiter()` call.
> 
> You can read more about what's supported and not supported with `ValueTask<T>` in https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/.

So.  Just as an epilogue, I am very grateful to have gotten that expert read from you.  

I ended up implementing the 'advanced pattern for the very hot codepath' as described in this link.  The performance improvement was dramatic.  I feel educated.  Thanks again!

**Comment by stephentoub** [comment]
Thanks for the nice note.  Very glad to hear the improvements were significant.

**Comment by nmoschkin** [comment]
> Thanks for the nice note. Very glad to hear the improvements were significant.

Yes.  I'm implementing the Level 3 market feed coding pattern for a library that I'm developing, and am eventually going to branch out.  So that exact path is what is up.

I actually profiled deserialization with string comparison versus deserialization with CRC-32 hash comparison.  and it was 600 ticks for string comparison in the JSON objects, and 200 ticks for just adding every byte I receive to a rolling CRC-32 hash and then just comparing them against the known hashes for the string variables in a JSON object.  So, obviously I kept that implementation.

This websocket pattern was the last key to the puzzle.  

https://github.com/nmoschkin/KuCoin.NET  

Specifically the file KuCoin.NET/KuCoin.NET/Websockets/Public/Level3Direct.cs
